---
title: "Using R for Computational Biology"
author: "Marc, Bronson, Debbie and Sean"
date: "3/9/2016"
output: html_document
---

This week we will start by introducing you to basic plotting functionalities and then move on to some of the ways that the R programming language intersects with computational Biology.

# Basic plotting.

One of the most important things that you can do to help you understand data is to visualize it.  And R is especially famous for providing many flexible systems for visualizing complex data.  Here in this course we will focus on the most basic of these systems in order to give people exposure to the fundamental graphics systems.  But students who are interested are strongly encouraged to also look at the `lattice` and `ggplot2` packages which each extend the core graphics capabilities in remarkable ways.

Probably the most basic function for plotting in R is just `plot`.  You can call `plot` just by passing it some data.  For example, here is some data:

```{r,carData}
head(cars)
```

And here is one way you can quickly plot it:

```{r,plot}
plot(cars)
```

Of course even just basic `plot` is a pretty capable method which handles a lot of different kinds of data.  In fact, you can get many different kind of plots output from `plot` just by changing both the object you pass in to it as well as the other arguments that you use.  So for example, the above can changed to a very different view (even with the same data), by just using the 'type' argument like so:

```{r,ploth}
plot(cars, type='h')
```


## some popular basic plot types (tables)

And of course R has many, many other handy plotting functions baked in. Below are some tables to summarize a few of the more popular options.

### Some popular basic plots for 1D data:

Function  |  Data    |  Plot
----------|----------|-----------------------
plot      | Numeric  | Scatterplot
barplot   | Numeric  | Barplot
boxplot   | Numeric  | Box and whiskers plot
hist      | Numeric  | Histogram


### Some popular basic plots for 2D data:

Function        |  Data            |  Plot
----------------|------------------|-----------------------
plot            | Numeric, Numeric | Scatterplot
plot            | Factor, Numeric  | Boxplots
smoothscatter   | Numeric, Numeric | Smooth scatterplot
boxplot         | list of Numeric  | Boxplots
barplots        | Matrix           | Barplots (side by side)



## saving plots out to file

A common thing that happens is that you will get a plot looking the way you like in R and then you may discover that you want to save that output to a file.  The thing to understand about how R handles plots is that when you make a plot you are sending the image to the X11 renderer (printing it to the panel on your screen).  So to save it, you really want to print it instead to a file on your filesystem.  To do that you want to call the appropriate function like `jpeg`, `png` or `pdf` which will then open a connection to a file (with a path that you specify) so that R can print the image out to the filesystem INSTEAD of to the plotting window.  Then if you plot your function again it will instead be sent to your file. And finally, by then calling the `dev.off` function you return R to the default state so that future plots will once again be sent to the plotting window on your screen.  So the code to "save a plot"" ends up looking like this example here:

```{r, saveplot, eval=FALSE}
jpeg(file = "myPlot.jpg")
plot(cars)
dev.off()
```

If that all of that seems a little weird to you, you can also just click on the export button from your Plots panel in RStudio.  But of course that option will only work from within RStudio, and so if you need to save a plot as an output from a script you will be out of luck...


***
####  <span style="color:blue">__Exercise R1:__</span> 


***





# Bioconductor and BiocViews

Now that you know the basics of R, you are ready to explore some of the different software packages that people have built using that foundation.  There are many thousands of open source packages freely available on the internet at places like CRAN.  But in addition to these resources Bioconductor has over a thousand packages that all relate to computational biology.  These packages have all been vetted, and are built and checked nightly to make sure they still function as expected.  Also: these packages all have vignettes to describe how to use them and have been labeled with special controlled vocabulary in order to make them easy to discover.  The special vocabulary is called 'biocviews' and you can see how the packages are organized according to the 'biocviews' terms on the [website](http://bioconductor.org/packages/release/BiocViews.html#___Software).  Searching through the site, you can find many popular packages that help with flow cytometry analysis, microarrays, RNA-seq, chip-seq, high throughput mass spectrometry to name just a few topics.

Once you have a package that interests you, you should go to it's landing page and look at the instructional vignettes.  And if you then discover that you have questions about how best to use the package you can search or post questions on the sites [help forum](https://support.bioconductor.org/).  

Finally many of the most popular packages are also described in the context of popular [workflows](http://bioconductor.org/help/workflows/), which show a high level view for how popular packages are often used together.  


# Common Bioconductor objects

There is not time in this course to introduce you to all of the packages in Bioconductor.  What we have tried to do instead, is teach you enough basic R so that this world of software will be an open door for you to walk into when you need to.  But while we won't have time to teach you about all 1000+ packages in the project, we can show you a few things that many of the most popular pacakges has in common.  And one of the things that holds Bioconductor together as a project are a shared set of objects that allow data to be easily moved between packages that often do very different kinds of analysis.  A couple of the most common object types are `ExpressionSet` objects and `GRanges` objects.  There are of course many many other objects that are derived from these objects and which are different in subtle ways, but if you have seen these objects, then those other objects should seem more familiar.

## `ExpressionSet`

`ExpressionSet` objects were initially developed to hold data from microarray experiments.  Essentially, they held a square of measured expression values where each column corresponded to another experiment and they also had space for tracking both annotation and phenotype data that related to this.  Here is an example that has data for 26 samples of microarray from the hgu95av2 platform from Affymetrix:

```{r, ExpressionSetSetup, echo=FALSE}
suppressMessages(library(Biobase))
```

```{r, ExpressionSet}
library(Biobase)
data(sample.ExpressionSet)
sample.ExpressionSet
```

### Why this is popular:

This is very convenient because having this object not only puts all the data in one place, but it allows for convenient subsetting of the data (keeping all of the relevant details in sync with each other as you do so).  It also creates a kind of shared currency for all packages that want to use this kind of data.  Now anyone who wants to do something creative with data of this kind has a convenient way to share their data.  This ultimately allows packages written by people in different corners of the globe be able to communicate.

As a result of it's popularity, this container has been adapted many times to a whole range of different data types.  One of the more recent popular examples is the `RangedSummmarizedExperiment` objects defined in the `SummarizedExperiment` package.  These objects are similar in design to the class `ExpressionSet` objects, but they have been built instead with high throughput sequencing data in mind.  `RangedSummmarizedExperiment` objects are still an area of active development, but you can see a vignette detailing their current status [here](http://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html).

Here is an example of a `RangedSummmarizedExperiment` object:

```{r, RangedSummmarizedExperiment}
data(airway, package="airway")
airway
```


## `GRanges` and `GRangesList`

Another popular type of object that you will see used in Bioconductor are range based objects.  These are predominantly used for high throughput sequencing data.  They are used here as a convenient way to represent data that is mapped onto a genome.  As a result, there are certain minimal things that all `GRanges` objects have to have baked in in order to effectively be 'placed' on a genome.  To get an idea about what needs to be in a a `GRanges` object, lets build one:

```{r, Granges}
gr <- GRanges(Rle(c("chr2", "chr2", "chr1", "chr3"), c(1, 3, 2, 4)),
               IRanges(1:10, width=10:1))
gr
```

As you can see `GRanges` objects have (at a minimum) all the information needed to represent specific ranges on the genome.  This is very convenient because most of the operations that people want to perform for genomic features involve rapidly comparing whether or not certain features overlap or not, and currently most genomic data is mapped relative to a common set of references.

In addition to `GRanges`, Sometimes it is helpful to create lists of `GRanges` objects in order to represent nested structures (like exons contained inside of transcripts).  For these circumstances there is the `GRangesList` object.  Usually you will see these objects when looking at certain kinds of annotations.  Here  is an example of how that can come up:

```{r, GRangesList}
library(Homo.sapiens)
txby <- transcriptsBy(Homo.sapiens, by='gene', columns=c('SYMBOL','GENEID'))
class(txby)
txby[3:4]
```


## `DataFrame` Objects

Another common object that you will encounter are `DataFrame` objects.  The `DataFrame` was created because many ranges in computational biology need to be labeled by rich annotations that have a many to one relationship with a given range.  For example consider what happens if we also request UNIPROT IDs to go with our transcript ranges in the previous example:

```{r, GRangesDataFrame}
library(Homo.sapiens)
txby <- transcriptsBy(Homo.sapiens, by='gene', columns=c('SYMBOL','GENEID','UNIPROT'))
txby[[1]]
```

Notice how there are multiple UNIPROT IDs associated with each gene?  If you then extract the metadata like this:

```{r, DataFrame}
meta <- mcols(txby[[1]])
meta
```

Then you can see that the GRanges object returned by the `transcriptsBy` method contains a `DataFrame` object.  This `DataFrame` object is returned when you call the `mcols` method and the `DataFrame` object allows the GRanges object that it is embedded into to store multiple values for the UNIPROT IDs associated with each gene.


## How can I learn more about what I might do with all these fancy S4 Objects?

Often you will encounter a new object and wonder: "what is that for"?  It's a perfectly reasonable question to find yourself asking.  And there is a good way  to get answers.  The thing to do in this circumstance is to call `methods` and then look at the list of methods associated with that object.  So for example, you can use `methods` to get a list of all the methods associated with `ExpressionSet` objects like this:

```{r, methods}
methods(class='ExpressionSet')
```

Now that you know the names of the associated methods, you can look up manual pages and explore more about what is possible for this object.  This is especially helpful when using Bioconductor where there are a lot of these rich S4 objects (and associated methods).


***
####  <span style="color:blue">__Exercise R2:__</span> 


***



# Literate programming.

Literate programming is the idea that a single document can contain both human ideas and executable machine code.  It was initially put forward by the legendary Donald Knuth.  And it is especially valuable in fields like science and mathematics where ideas need to be reproducible and testable when shared with colleagues.  Over the years there have been several attempts to make this workable and they have seen varying degrees of success, but one of the most usable and successful applications of this is the R markdown (.Rmd) document.

## What is an .Rmd file?

An R markdown (.Rmd) document allows you to embed R code right into the middle of your prose.  You can make one right from your RStudio session by simply choosing it from the File menu.  The markdown is extremely simple and easy to learn.  But for a fully detailed description of all advanced features, be sure to have a look at this document here: 
```{bash, eval=FALSE}
http://rmarkdown.rstudio.com/
```

## Reproducible research

But why would you want to do this?  Probably the best reason is that you want to know both what code you wrote and why you wrote it later on.  By documenting your code with .Rmd files, you can come back later and see both what you were thinking as well as what code you ran.  It is also a nicely explicit way of sharing workflows with colleagues who otherwise might struggle to reproduce your work.

As an example of a use case for .Rmd, all the handouts for this course were created with .Rmd documents.  Using .Rmd allowed us to run both a spelling check __and__ a code check to make sure that all the code blocks ran as expected before we went to press.

Also useful to know is that you can use .Rmd for languages other than just R.  Sure it's great for R, but you can also write code chunks that run a whole array of alternative language calls.  Have a look at this list of alternate possibilities:

```{r, echo=FALSE}
names(knitr::knit_engines$get())
```
Lets do a quick exercise to learn what markdown looks like and how you can edit it

***
####  <span style="color:blue">__Exercise R3:__</span> 


***


## Shiny

Now that we have talked about both basic plotting and R markdown lets also briefly mention the `shiny` package.  The `shiny` package was created to bring the interactivity of javascript and connect it with R's famous plotting abilities.  The result is a system that lets you take your data and make your plots interactive.  If you are curious to explore shiny, go to the 'File' menu, choose 'New File', and then 'R markdown' and then choose 'Shiny' on the left hand side of the pop up window.




# Version control.

So writing your code as .Rmd files will go a long ways towards helping you compose your ideas so that things can be tested easily.  But what if you have a lot of R code and you are making a lot of changes?  How will you keep track of what has changed?  And if you work on a piece of code with a co-worker?  How will you both be able to work on it without losing your minds?

The answer is to use version control software.  Version control software is something that all software developers use, and it's very much worth your time to learn to use it too.  It's not difficult to learn, and if you make good use of it, it can not only save your bacon, but it can make your life a lot easier as well.  There have been many popular flavors of version control software that have come out over the years.  A few you might have heard about are `rcs`, `cvs`, `svn` and more recently `git`. 

Version control is probably too big a topic to do justice to in a beginner course (it could be a short class on it's own).  But it is mentioned here so that you know it is something that you should really learn about.


***

# Homework exercises.

Some of you have been asking for homework.  One great activity is to work through external vignettes.  Some good packages to look at ahead of next week are these:

```
http://bioconductor.org/packages/release/bioc/html/ShortRead.html

http://bioconductor.org/packages/release/bioc/html/DESeq2.html

http://bioconductor.org/packages/release/bioc/html/AnnotationDbi.html

http://bioconductor.org/packages/release/bioc/html/GenomicFeatures.html
```

If you look at the .pdf files for these packages (even if you don't fully understand them) you will be helping to set yourself up for success next time.  Another thing to look is to explore the Bioconductor packages here:

```
http://bioconductor.org/packages/release/BiocViews.html#___Software
```

As you do that, pay close attention to the tree hierarchy and to the search boxes on that page.  These all allow you to search the repository in different ways until you can find a package written to solve a problems that you are working on.  Once you find software written to solve problems you are working on, you should read through the vignettes for it.



***
***
***

## Answers For Exercises:


####  <span style="color:blue">__Exercise R1:__</span> 

Basic questions about plotting (plot some data, change some labels, save it)

Look at the `mtcars` data.frame.  Now make a scatterplot to see if heavier cars tend to have more horsepower. 

Now make a `barplot` to look at the average weight for each number of cylinders in the table.  To do this, use `split` and `mean` to make a vector that summarizes the average weight per number of cylinders and then make a `barplot`

Now save your barplot to a pdf file.


#####  <span style="color:blue">__Answer for Exercise R1:__</span> 

```{r,ans1}
plot(mtcars[,c('wt','hp')])

```

```{r,ans1b}
wts <- split(mtcars[,c('wt')],f=mtcars[,c('cyl')])
wts <- unlist(lapply(wts, mean))
barplot(wts)
```

```{r,ans1c}
jpeg(file = "myBarPlot.jpg")
barplot(wts)
dev.off()
```


####  <span style="color:blue">__Exercise R2:__</span> 

Basic questions about Bioconductor objects (use methods on 'OrganismDb' (which was introduced but not explained). Have them look up the help for select() and then use Homo.sapiens to extract information about it)

#####  <span style="color:blue">__Answer for Exercise R2:__</span> 

```{r,ans2}

```



####  <span style="color:blue">__Exercise R3:__</span> 

Lets start by using RStudio to create an .Rmd file.  You can do this by going to the `File` menu, and selecting `New File` followed by `R Markdown ...`.  Next you will notice that you can create many different 'flavors' of .Rmd file, for our example, just choose a basic 'Document', and fill in the title and authors.  

Now look at the document that is opened up in RStudio.  Notice how certain sections of it are marked by triple back ticks (```).  These are code chunks which will be evaluated when the document is rendered.  You can put whatever R code you want to into these chunks and you can surround it be whatever prose you need to include in the document.

Now add a new code chunk to the document.

#####  <span style="color:blue">__Answer for Exercise R3:__</span> 


The code chunk that you insert into your document should look like this (minus the "|" symbol on the left hand edge of the following display).  The reason for the "|" is because it is hard to get RStudio to render what a code chunk should look like since it wants to interpret that code chunk...

````
|       ```{r}
        myPasteFun <- function(x){
          paste(x, 'extension', sep=".")
        }
        myPasteFun('testStr')
        ```
````



